modules/heat_conduction/src/userobjects/ViewFactorRayStudy.C

```diff
@@ -0,0 +1,554 @@
+  for (const auto & start_elem : _start_elems)
+  {
+    num_local_start_points += start_elem._points.size();
+    num_local_rays += start_elem._points.size() * _num_dir;
+  }
+
+  // Print out totals while we're here
+  std::size_t num_total_points = num_local_start_points;
+  std::size_t num_total_rays = num_local_rays;
+  _communicator.sum(num_total_points);
+  _communicator.sum(num_total_rays);
+  _console << "ViewFactorRayStudy generated " << num_total_points
+           << " points with an angular quadrature of " << _num_dir
+           << " directions per point requiring " << num_total_rays << " rays" << std::endl;
+
+  // Reserve space in the buffer ahead of time before we fill it
+  reserveRayBuffer(num_local_rays);
+
+  Point direction;
+
+  // loop through all starting points and spawn rays from each for each point and angle
+  for (const auto & start_elem : _start_elems)
+  {
+    // Get normal for the element we're starting on
+    auto inward_normal =
+        getSideNormal(start_elem._start_elem, start_elem._incoming_side, /* tid = */ 0);
+    // We actually want the normal of the original element (remember that we may swap starting
+    // elements to the element on the other face per requirements of the ray tracer)
+    if (start_elem._start_elem != start_elem._elem)
+      inward_normal *= -1;
+    // Lastly, if the boundary is external and the internal convention is positive, we must
+    // switch the normal because our AQ uses the inward normal
+    if (_internal_convention == 0 &&
+        !start_elem._start_elem->neighbor_ptr(start_elem._incoming_side))
+      inward_normal *= -1;
+
+    // Rotation for the quadrature to align with the normal; in 3D we can do all of this
+    // once up front using the 3D aq object. For 2D, we will do it within the direction loop
+    if (_is_3d)
+      _3d_aq->rotate(inward_normal);
+
+    // Loop through all points and then all directions
+    for (std::size_t start_i = 0; start_i < start_elem._points.size(); ++start_i)
+      for (std::size_t l = 0; l < _num_dir; ++l)
+      {
+        // Get direction of the ray; in 3D we already rotated, in 2D we rotate here
+        if (_is_3d)
+          direction = _3d_aq->getDirection(l);
+        else
+        {
+          const Real sin_theta = std::sin(_2d_aq_angles[l]);
+          const Real cos_theta = std::cos(_2d_aq_angles[l]);
+          direction(0) = cos_theta * inward_normal(0) - sin_theta * inward_normal(1);
+          direction(1) = sin_theta * inward_normal(0) + cos_theta * inward_normal(1);
+          direction(2) = 0;
+        }
+
+        // Angular weight function differs in 2D/3D
+        // 2D: the quadrature abscissae are the angles between direction & normal.
+        //     The integrand is the cosine of that angle
+        // 3D: the quadrature abscissae are the azimuthal angle phi and the cosine of the angle
+        //     between normal and direction (== mu). The weight is mu in that case.
```

Comment:
why does phi come in the weight again? inward_normal * direction only has the cosine right 
does it comes from _3d_q->getTotalWeight(l) ?
```suggestion
        //     between normal and direction (= mu). The weight is mu in that case.
```

Reply:
The wording here is bad. The integrand is mu, not the weight.

Reply:
done